numpy/lib/function_base.py

```diff
@@ -3929,21 +3929,33 @@ def _quantile_ureduce_func(a, q, axis=None, out=None, overwrite_input=False,
             indices_above = indices_above[:-1]
             n = np.isnan(ap[-1:, ...])
 
-        x1 = take(ap, indices_below, axis=axis) * weights_below
-        x2 = take(ap, indices_above, axis=axis) * weights_above
+        x1 = take(ap, indices_below, axis=axis)
+        x2 = take(ap, indices_above, axis=axis)
+
+        diff_x2_x1 = x2 - x1
+        r_above = x1 + diff_x2_x1 * weights_above
+        r_below = x2 - diff_x2_x1 * weights_below
 
         # ensure axis with q-th is first
-        x1 = np.moveaxis(x1, axis, 0)
-        x2 = np.moveaxis(x2, axis, 0)
+        r_above = np.moveaxis(r_above, axis, 0)
+        r_below = np.moveaxis(r_below, axis, 0)
 
         if zerod:
-            x1 = x1.squeeze(0)
-            x2 = x2.squeeze(0)
+            weights_above = weights_above.squeeze(0)
+            weights_below = weights_below.squeeze(0)
+            r_above = r_above.squeeze(0)
+            r_below = r_below.squeeze(0)
+
+        r = np.where(weights_above < 0.5, r_above, r_below)
+
+        if r.ndim == 0:
+            # get the element stored in the 0-d array for backward
+            # compatibility
+            r = r[()]
```

Comment:
`add(x1, x2)` would have return a scalar so it seems that `r[()]` is doing the job. @eric-wieser is there something more appropriate regarding this case?

Reply:
This doesn't quite belong here, it should only happen if `out is None`, as an `else` in the `if` below.